src/python/pants/backend/python/goals/setup_py.py

```diff
@@ -563,72 +560,46 @@ async def generate_chroot(request: SetupPyChrootRequest) -> SetupPyChroot:
-                f"`entry_point='{entry_point.module}:main'. See {url}."
-            )
-        entry_point_requests.append(ResolvePexEntryPointRequest(binary[PexEntryPointField]))
-
-    binary_entry_points = await MultiGet(
-        Get(ResolvedPexEntryPoint, ResolvePexEntryPointRequest, request)
-        for request in entry_point_requests
-    )
-
-    entry_points_from_with_binaries = {
-        key: binary_entry_point.val.spec
-        for key, binary_entry_point in zip(key_to_binary_spec.keys(), binary_entry_points)
-        if binary_entry_point.val is not None
-    }
-
-    # Collect entry points from `python_distribution(entry_points=...)`
-    entry_points_from_field = {}
-    if exported_target.target.has_field(PythonDistributionEntryPointsField):
-        entry_points_field = await Get(
+    # Resolve entry points from python_distribution(entry_points=...) and from
+    # python_distribution(provides=setup_py(entry_points=...).with_binaries(...)
+    entry_points_field, provides_field = await MultiGet(
+        Get(
             ResolvedPythonDistributionEntryPoints,
             ResolvePythonDistributionEntryPointsRequest(
-                exported_target.target[PythonDistributionEntryPointsField]
+                entry_points_field=exported_target.target.get(PythonDistributionEntryPointsField)
             ),
-        )
+        ),
+        Get(
+            ResolvedPythonDistributionEntryPoints,
+            ResolvePythonDistributionEntryPointsRequest(
+                provides_field=exported_target.target.get(PythonProvidesField)
+            ),
+        ),
+    )
 
-        entry_points_from_field = {
-            category: {ep_name: ep_val.entry_point.spec for ep_name, ep_val in entry_points.items()}
-            for category, entry_points in entry_points_field.val.items()
-        }
+    entry_points_from_field = {
+        category: {ep_name: ep_val.entry_point.spec for ep_name, ep_val in entry_points.items()}
+        for category, entry_points in entry_points_field.val.items()
+    }
 
-    # Collect any entry points from the setup_py() object. Note that this was already normalized in
-    # `python_artifacts.py`.
-    entry_points_from_provides = setup_kwargs.get("entry_points", {})
+    entry_points_from_provides = {
+        category: {ep_name: ep_val.entry_point.spec for ep_name, ep_val in entry_points.items()}
+        for category, entry_points in provides_field.val.items()
+    }
 
     # Gather entry points with source description for any error messages when merging them.
     entry_point_sources = {
-        f"{exported_addr}'s field `provides=setup_py().with_binaries()`": {
-            "console_scripts": entry_points_from_with_binaries
-        },
         f"{exported_addr}'s field `entry_points`": entry_points_from_field,
-        f"{exported_addr}'s field `provides=setup_py(..., entry_points={...})`": entry_points_from_provides,
```

Comment:
The `with_binaries()` are now merged during resolve of the provides field, so we're left with the two resolves for each field, entry points and provides, respectively. The downside is, that when there is a conflict with a entry point from the distributions field, we can't say if it is with setup_py().with_binaries or with setup_py(entry_points) any more..

Reply:
That is totally reasonable, esp. becauase with_binaries() is now deprecated. Thanks for the disclaimer!